cpp/src/bitmask/valid_if.cuh

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */ 
+
+#ifndef __BITMASK_VALID_IF_CUH__
+#define __BITMASK_VALID_IF_CUH__
+
+#include <cudf/types.h>
+#include <utilities/error_utils.hpp>
+#include <utilities/bit_util.cuh>
+#include <utilities/cuda_utils.hpp>
+#include <bitmask/legacy/bit_mask.cuh>
+
+#include <cub/cub.cuh>
+
+using bit_mask::bit_mask_t;
+
+namespace cudf {
+
+namespace detail {
+
+constexpr int warp_size = 32;
+
+/**
+ * @brief for each warp in the block do a reduction (summation) of the
+ * `__popc(bit_mask)` on a certain lane (default is lane 0).
+ * @param[in] bit_mask The bit_mask to be reduced.
+ * @return[out] result of each block is returned in thread 0.
+ */
+template <class bit_container, int lane = 0>
+__device__ __inline__ gdf_size_type single_lane_popc_block_reduce(bit_container bit_mask) {
+  
+  static __shared__ gdf_size_type smem[warp_size];
+  
+  int lane_id = (threadIdx.x % warp_size);
+  int warp_id = (threadIdx.x / warp_size);
+
+  // Assuming one lane of each warp holds the value that we want to perform
+  // reduction
+  if (lane_id == lane) {
+    smem[warp_id] = __popc(bit_mask);
+  }
+  __syncthreads();
+
+  if (warp_id == 0) {
+    // Here I am assuming maximum block size is 1024 and 1024 / 32 = 32
+    // so one single warp is enough to do the reduction over different warps
+    bit_mask = (lane_id < (blockDim.x / warp_size)) ? smem[lane_id] : 0;
```

Comment:
Why is this called bit_mask? `smem[lane_id]` contains the result of __popc(), which is a count, not a mask.  I think you should call this `count`...
```suggestion
    gdf_size_type count = (lane_id < (blockDim.x / warp_size)) ? warp_counts[lane_id] : 0;
```